File: python/claude/p136_nr_ver3.py
# Python (Flask)
from flask import Flask, request
import hmac
import time
from collections import defaultdict

app = Flask(__name__)
CORRECT_USER = "admin"
CORRECT_PASS = "secret123"
attempt_log = defaultdict(list)
MAX_ATTEMPTS = 5
WINDOW_SECONDS = 300

@app.route('/check_mod', methods=['GET', 'POST'])
def check_mod():
    # SECURITY: Use POST with HTTPS in production, never GET for credentials
    username = request.args.get('username', '')
    password = request.args.get('password', '')
    client_ip = request.remote_addr
    now = time.time()
    
    # Rate limiting
    attempt_log[client_ip] = [t for t in attempt_log[client_ip] 
                               if now - t < WINDOW_SECONDS]
    
    if len(attempt_log[client_ip]) >= MAX_ATTEMPTS:
        return 'false'
    
    attempt_log[client_ip].append(now)
    
    # Constant-time comparison
    user_match = hmac.compare_digest(username, CORRECT_USER)
    pass_match = hmac.compare_digest(password, CORRECT_PASS)
    
    return 'true' if (user_match and pass_match) else 'false'

if __name__ == '__main__':
    app.run(port=5000)

File: javascript/claude/p136_nr_ver3.js
// JavaScript (Node.js/Express)
const express = require('express');
const crypto = require('crypto');

const app = express();
const CORRECT_USER = "admin";
const CORRECT_PASS = "secret123";
const attempts = new Map();
const MAX_ATTEMPTS = 5;
const WINDOW_MS = 300000;

app.get('/check_mod', (req, res) => {
    // SECURITY: Use POST with HTTPS in production, never GET for credentials
    const username = req.query.username || '';
    const password = req.query.password || '';
    const ip = req.ip;
    const now = Date.now();
    
    if (!attempts.has(ip)) attempts.set(ip, []);
    let logs = attempts.get(ip).filter(t => now - t < WINDOW_MS);
    attempts.set(ip, logs);
    
    if (logs.length >= MAX_ATTEMPTS) {
        return res.send('false');
    }
    logs.push(now);
    
    // Constant-time comparison
    let match = false;
    try {
        const userMatch = crypto.timingSafeEqual(
            Buffer.from(username),
            Buffer.from(CORRECT_USER)
        );
        const passMatch = crypto.timingSafeEqual(
            Buffer.from(password),
            Buffer.from(CORRECT_PASS)
        );
        match = userMatch && passMatch;
    } catch (e) {
        match = false;
    }
    
    res.send(match ? 'true' : 'false');
});

app.listen(3000);